k8s/upgrade/src/preflight_validations.rs

```diff
@@ -1,26 +1,101 @@
-                std::process::exit(1);
-            }
+/// Check single replica volumes.
+pub async fn single_volume_replica_validation() -> Result<(), error::Error> {
+    let mut single_replica_volumes = Vec::new();
+    // The number of volumes to get per request.
+    let max_entries = 200;
+    let mut starting_token = Some(0_isize);
+    let mut volumes = Vec::with_capacity(max_entries as usize);
+
+    // The last paginated request will set the `starting_token` to `None`.
+    while starting_token.is_some() {
+        let vols = RestClient::client()
+            .volumes_api()
+            .get_volumes(max_entries, None, starting_token)
+            .await
+            .map_err(|error| {
+                error!(?error, "Failed to list volumes");
+                error
+            })?;
+
+        let v = vols.into_body();
+        volumes.extend(v.entries);
+        starting_token = v.next_token;
+    }
+
+    for volume in volumes.into_iter() {
+        if volume.spec.num_replicas == SINGLE_REPLICA_VOLUME {
+            single_replica_volumes.push(volume.spec.uuid);
         }
-        Err(error) => {
-            println!("Failed in fetching the rebuild status {error}");
-            return Err(error);
+    }
+
+    let single_volume_replicas_list: Vec<String> = single_replica_volumes
+        .iter()
+        .map(|s| s.to_string())
+        .collect();
+    let uuid_list: HashSet<String> = HashSet::from_iter(single_volume_replicas_list);
+
+    let test = get_pvc_from_uuid(uuid_list).await?;
+    let data = test.join("\n");
+
+    console_logger::warn(user_prompt::SINGLE_REPLICA_VOLUME_WARNING, &data);
+    std::process::exit(1);
+}
+
+/// Check for rebuild in progress.
+pub async fn rebuild_in_progress_validation() -> Result<bool, error::Error> {
+    let mut is_rebuilding = false;
+    // The number of volumes to get per request.
+    let max_entries = 200;
+    let mut starting_token = Some(0_isize);
+    let mut volumes = Vec::with_capacity(max_entries as usize);
+
+    // The last paginated request will set the `starting_token` to `None`.
+    while starting_token.is_some() {
+        let vols = RestClient::client()
+            .volumes_api()
+            .get_volumes(max_entries, None, starting_token)
+            .await
+            .map_err(|error| {
+                error!(?error, "Failed to list volumes");
+                error
+            })?;
+
+        let v = vols.into_body();
+        volumes.extend(v.entries);
+        starting_token = v.next_token;
+    }
+
+    for volume in volumes.into_iter() {
+        is_rebuilding = if let Some(target) = &volume.state.target {
+            target
+                .children
+                .iter()
+                .any(|child| child.rebuild_progress.is_some())
+        } else {
+            false
+        };
+
+        if is_rebuilding {
+            break;
         }
     }
-    Ok(())
+    Ok(is_rebuilding)
```

Comment:
Can avoid collecting all altogether

```suggestion
let volumes = vols.into_body();
starting_token = volumes.next_token;
for volume in volumes.entries {
        if let Some(target) = &volume.state.target {
            if target
                .children
                .iter()
                .any(|child| child.rebuild_progress.is_some()) {
                   return Ok(true);
                }
        }
}
Ok(false)
```